### ISSUE TITLE
cannot install local dependencies with prepublish scripts

### ISSUE THREAD

**Issue by timoxley** [npm/npm#7040]
### tldr;

prepublish scripts always run for local dependencies on `npm install`, even when using `npm install --production`.

Consider this `package.json`

``` json
{
  "name": "npm-issue-local-dependency-prepublish",
  "version": "0.0.0",
  "dependencies": {
    "local-pkg": "file:local-pkg"
  }
}

```

And this `./local-pkg/package.json`

``` json
{
  "name": "local-pkg",
  "version": "1.0.0",
  "scripts": {
    "prepublish": "tap --help"
  },
  "dependencies": {
    "tap": "^0.4.13"
  }
}
```

Ensuring there is no `tap` installed on the PATH, and all `node_modules` are clean:

```
npm uninstall -g tap
rm -rf node_modules
rm -rf local-pkg/node_modules
```

When running `npm install` from the first package, it will produce the following error:

```
> local-pkg@1.0.0 prepublish /Users/timoxley/Projects/tests/npm-issue-local-dependency-prepublish/local-pkg
> tap --help

sh: tap: command not found

npm ERR! addLocalDirectory Could not pack "/Users/timoxley/Projects/tests/npm-issue-local-dependency-prepublish/local-pkg" to "/Users/timoxley/.npm/local-pkg/1.0.0/package.tgz"
npm ERR! addLocal Could not install /Users/timoxley/Projects/tests/npm-issue-local-dependency-prepublish/local-pkg
npm ERR! Darwin 14.1.0
npm ERR! argv "node" "/usr/local/bin/npm" "install"
npm ERR! node v0.10.35
npm ERR! npm  v2.1.16
npm ERR! code ELIFECYCLE
npm ERR! local-pkg@1.0.0 prepublish: `tap --help`
npm ERR! Exit status 127
npm ERR!
npm ERR! Failed at the local-pkg@1.0.0 prepublish script 'tap --help'.
npm ERR! This is most likely a problem with the local-pkg package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     tap --help
npm ERR! You can get their info via:
npm ERR!     npm owner ls local-pkg
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/timoxley/Projects/tests/npm-issue-local-dependency-prepublish/npm-debug.log
```

**I believe this is incorrect behaviour.** Please correct me if I'm wrong.

The original change that triggers this behaviour was introduced here:
https://github.com/npm/npm/commit/2389525a1df6d20e780cca5887f1b73ff583ce8d
https://github.com/npm/npm/issues/3854

**Comment by timoxley** [comment]
Not sure solution here, perhaps a check if it's a local dependency somewhere here:

[![image](https://cloud.githubusercontent.com/assets/43438/5593118/3ae4536e-9242-11e4-9c0a-e7922ed0233c.png)](https://github.com/npm/npm/blob/7a22d0ee18ddd093e233f8d4ed26fa0b1a8235ec/lib/cache/add-local.js#L93-L94)

https://github.com/npm/npm/blob/7a22d0ee18ddd093e233f8d4ed26fa0b1a8235ec/lib/cache/add-local.js#L93-L94

or the prepublish could be triggered in the `npm pack` command itself rather than inside in the pack library code.

**Comment by othiym23** [comment]
This looks like a legit bug, but I am up to my eyeballs in alligators right now, and @iarna is trying to shove `npm@3` over the finishing line, so I'm afraid we won't have the spare cycles to get to this until a few weeks have passed.

**Comment by boneskull** [comment]
This bites me all the time.  Many packages have `prepublish` scripts, and I install them via `npm` and thereafter cannot `npm pack` for all sorts of reasons--mainly either files being excluded within the `.npmignore`, or global scripts not installed.

**Comment by diwo** [comment]
I'm running into the same issue when installing a local dependency that has a prepublish step which depends on a dev dependency. The prepublish script compiles non-javascript source files into javascript to be consumed. When I install another package that depends on this local dependency, the command fails on the prepublish step of the local dependency due to its dev dependency not being present.

In my case however, I need the prepublish script to be executed so that the compiled javascript from the local dependency can be loaded with ```require```.

So instead of excluding the prepublish step from being run on a local dependency, perhaps the solution could be to run install on the local dependency before running prepublish and pack.

**Comment by boneskull** [comment]
@timoxley Is the bug here that `npm install --production` runs `prepublish` anyway?

**Comment by diwo** [comment]
@boneskull I think that's covered in https://github.com/npm/npm/pull/4461

**Comment by iarna** [comment]
IMO the bug here is that npm should be running install on local deps before trying to pack them, as prepublish scripts are often not optional.

**Comment by boneskull** [comment]
@iarna But if you installed the package with the `--production` flag, I wouldn't expect any dev dependencies to be installed, and often `prepublish` relies on those.

My use case is this:

- `npm install --production foo`
- `cd foo`
- `npm pack`
- ERROR: foo's `prepublish` script tried to use a dev dependency

If #4461 was closed, then I'm not sure what the fix is other than don't run `prepublish` before packing.  I don't see the point of running `prepublish` before packing (but I'm probably missing something) if you're just going to `npm install` the tarball later.

**Comment by boneskull** [comment]
I should note that I *don't* want to pack up the dev dependencies, because I installed with `--production`.  My terrible workaround is to manually prune `prepublish` from `package.json` before packing the problem package.  :smile:

**Comment by diwo** [comment]
```npm pack``` doesn't package dependencies/devDependencies into the tarball so that shouldn't be a problem

**Comment by iarna** [comment]
@boneskull As @diwo says, it would install them for the purposes of building a tarball to add to the cache. They would not end up in your application.

**Comment by boneskull** [comment]
@iarna right, brain fart.  I think you're right then; `pack` should run `install`.

**Comment by petebrowne** [comment]
I'm running into this issue. What's the status of this? Are there any workarounds?
